File: evals/python/src/z3rno_evals/runner.py
"""Runs the golden-dataset eval against a live `z3rno.Client` and produces
a machine-readable `results.json` plus a human-readable `report.md`.
"""

import json
import math
import time
from pathlib import Path
from typing import Any

from .embed import hash_embed
from .metrics import LatencyPercentiles, faithfulness, latency_percentiles, mrr, recall_at_k


def load_fixture(path: Path) -> dict[str, Any]:
    with open(path, encoding="utf-8") as f:
        return json.load(f)


def _mean_skip_nan(values: list[float]) -> float:
    clean = [v for v in values if not math.isnan(v)]
    if not clean:
        return float("nan")
    return sum(clean) / len(clean)


def run_eval(client: Any, fixture: dict[str, Any]) -> dict[str, Any]:
    """Seeds `fixture["seed"]` into `client`, runs `fixture["items"]` as
    recall queries, and returns a results dict with per-item scores and
    aggregates. `client` is a `z3rno.Client` (or anything matching its
    store/recall interface).
    """
    tenant_id = fixture["tenant_id"]

    # Seed every memory, remembering fixture id -> real store()-returned id.
    seed_id_map: dict[str, str] = {}
    for record in fixture["seed"]:
        embedding = hash_embed(record["content"])
        memory = client.store(
            tenant_id=tenant_id,
            tier=record["tier"],
            content=record["content"],
            embedding=embedding,
            metadata=record.get("metadata") or {},
        )
        seed_id_map[record["id"]] = memory["id"]

    per_item: list[dict[str, Any]] = []
    latencies_ms: list[float] = []

    for item in fixture["items"]:
        top_k = item.get("top_k", 5)
        query_vec = hash_embed(item["query"])

        start = time.perf_counter()
        retrieved = client.recall(tenant_id=tenant_id, query=query_vec, k=top_k)
        latency_ms = (time.perf_counter() - start) * 1000.0
        latencies_ms.append(latency_ms)

        retrieved_ids = [m["id"] for m in retrieved]
        expected_ids = [seed_id_map[sid] for sid in item["expected_seed_ids"]]

        recalled_context = " ".join(m["content"] for m in retrieved)

        r_at_k = recall_at_k(retrieved_ids, expected_ids, top_k)
        r_mrr = mrr(retrieved_ids, expected_ids)
        r_faith = faithfulness(item.get("expected_answer", ""), recalled_context)

        per_item.append(
            {
                "id": item["id"],
                "query": item["query"],
                "tags": item.get("tags", []),
                "recall_at_k": r_at_k,
                "mrr": r_mrr,
                "faithfulness": r_faith,
                "latency_ms": latency_ms,
                "retrieved_ids": retrieved_ids,
                "expected_ids": expected_ids,
            }
        )

    aggregates = {
        "recall_at_k_mean": _mean_skip_nan([i["recall_at_k"] for i in per_item]),
        "mrr_mean": _mean_skip_nan([i["mrr"] for i in per_item]),
        "faithfulness_mean": _mean_skip_nan([i["faithfulness"] for i in per_item]),
        "latency": _latency_dict(latency_percentiles(latencies_ms)),
        "item_count": len(per_item),
    }

    return {
        "dataset": fixture.get("name", "unknown"),
        "tenant_id": tenant_id,
        "items": per_item,
        "aggregates": aggregates,
    }


def _latency_dict(p: LatencyPercentiles) -> dict[str, float]:
    return {"p50": p.p50, "p95": p.p95, "p99": p.p99, "mean": p.mean, "count": p.count}


def write_results(results: dict[str, Any], out_dir: Path) -> tuple[Path, Path]:
    out_dir.mkdir(parents=True, exist_ok=True)

    results_path = out_dir / "results.json"
    results_path.write_text(json.dumps(results, indent=2), encoding="utf-8")

    report_path = out_dir / "report.md"
    report_path.write_text(_render_report_md(results), encoding="utf-8")

    return results_path, report_path


def _render_report_md(results: dict[str, Any]) -> str:
    agg = results["aggregates"]
    lat = agg["latency"]
    lines = [
        f"# z3rno-evals report — {results['dataset']}",
        "",
        "## Aggregates",
        "",
        f"- recall@k (mean): {agg['recall_at_k_mean']:.4f}",
        f"- MRR (mean): {agg['mrr_mean']:.4f}",
        f"- faithfulness (mean): {agg['faithfulness_mean']:.4f}",
        f"- latency p50/p95/p99 (ms): {lat['p50']:.2f} / {lat['p95']:.2f} / {lat['p99']:.2f}",
        f"- latency mean (ms): {lat['mean']:.2f}",
        f"- items: {agg['item_count']}",
        "",
        "## Per-item",
        "",
        "| id | recall@k | mrr | faithfulness | latency_ms | tags |",
        "| --- | --- | --- | --- | --- | --- |",
    ]
    for item in results["items"]:
        lines.append(
            f"| {item['id']} | {item['recall_at_k']:.3f} | {item['mrr']:.3f} "
            f"| {item['faithfulness']:.3f} | {item['latency_ms']:.2f} | {', '.join(item['tags'])} |"
        )
    lines.append("")
    return "\n".join(lines)

File: evals/typescript/src/runner.ts
import * as fs from "node:fs";
import * as path from "node:path";
import type { Client, MemoryDto } from "@z3rno/sdk";

import { hashEmbed } from "./embed";
import { faithfulness, latencyPercentiles, mrr, recallAtK, type LatencyPercentiles } from "./metrics";
import type { GoldenFixture } from "./fixture";

export interface ItemResult {
  id: string;
  query: string;
  tags: string[];
  retrievedIds: string[];
  expectedIds: string[];
  recallAtK: number;
  mrr: number;
  faithfulness: number;
  latencyMs: number;
}

export interface EvalReport {
  fixtureName: string;
  tenantId: string;
  seedCount: number;
  itemCount: number;
  items: ItemResult[];
  aggregate: {
    meanRecallAtK: number;
    meanMrr: number;
    meanFaithfulness: number;
    latency: LatencyPercentiles;
  };
}

function mean(values: number[]): number {
  const finite = values.filter((v) => !Number.isNaN(v));
  if (finite.length === 0) return NaN;
  return finite.reduce((s, v) => s + v, 0) / finite.length;
}

/**
 * Seeds every `fixture.seed` record, then runs every `fixture.items` query
 * against the recalled results, scoring recall@k/MRR/faithfulness per item
 * plus aggregate latency percentiles.
 */
export async function runEval(client: Client, fixture: GoldenFixture): Promise<EvalReport> {
  const seedIdMap = new Map<string, string>(); // fixture seed id -> real store()-returned id

  for (const seed of fixture.seed) {
    const stored = await client.store({
      tenantId: fixture.tenant_id,
      tier: seed.tier,
      content: seed.content,
      embedding: hashEmbed(seed.content),
      metadata: seed.metadata,
      links: [],
    });
    seedIdMap.set(seed.id, stored.id);
  }

  const items: ItemResult[] = [];
  const latencies: number[] = [];

  for (const item of fixture.items) {
    const start = performance.now();
    const recalled: MemoryDto[] = await client.recall({
      tenantId: fixture.tenant_id,
      query: hashEmbed(item.query),
      k: item.top_k,
    });
    const latencyMs = performance.now() - start;
    latencies.push(latencyMs);

    const retrievedIds = recalled.map((m) => m.id);
    const expectedIds = item.expected_seed_ids
      .map((fixtureId) => seedIdMap.get(fixtureId))
      .filter((id): id is string => id !== undefined);
    const recalledContext = recalled.map((m) => m.content).join(" ");

    items.push({
      id: item.id,
      query: item.query,
      tags: item.tags,
      retrievedIds,
      expectedIds,
      recallAtK: recallAtK(retrievedIds, expectedIds, item.top_k),
      mrr: mrr(retrievedIds, expectedIds),
      faithfulness: faithfulness(item.expected_answer, recalledContext),
      latencyMs,
    });
  }

  return {
    fixtureName: fixture.name,
    tenantId: fixture.tenant_id,
    seedCount: fixture.seed.length,
    itemCount: fixture.items.length,
    items,
    aggregate: {
      meanRecallAtK: mean(items.map((i) => i.recallAtK)),
      meanMrr: mean(items.map((i) => i.mrr)),
      meanFaithfulness: mean(items.map((i) => i.faithfulness)),
      latency: latencyPercentiles(latencies),
    },
  };
}

export function loadFixture(fixturePath: string): GoldenFixture {
  return JSON.parse(fs.readFileSync(fixturePath, "utf-8")) as GoldenFixture;
}

function reportMarkdown(report: EvalReport): string {
  const lines: string[] = [];
  lines.push(`# z3rno TypeScript eval report — ${report.fixtureName}`);
  lines.push("");
  lines.push(`Tenant: \`${report.tenantId}\` · seeded ${report.seedCount} memories · ${report.itemCount} queries`);
  lines.push("");
  lines.push("## Aggregate scores");
  lines.push("");
  lines.push("| metric | value |");
  lines.push("| --- | --- |");
  lines.push(`| mean recall@k | ${report.aggregate.meanRecallAtK.toFixed(4)} |`);
  lines.push(`| mean MRR | ${report.aggregate.meanMrr.toFixed(4)} |`);
  lines.push(`| mean faithfulness | ${report.aggregate.meanFaithfulness.toFixed(4)} |`);
  lines.push(`| latency p50 (ms) | ${report.aggregate.latency.p50.toFixed(2)} |`);
  lines.push(`| latency p95 (ms) | ${report.aggregate.latency.p95.toFixed(2)} |`);
  lines.push(`| latency p99 (ms) | ${report.aggregate.latency.p99.toFixed(2)} |`);
  lines.push(`| latency mean (ms) | ${report.aggregate.latency.mean.toFixed(2)} |`);
  lines.push("");
  lines.push("## Per-item results");
  lines.push("");
  lines.push("| id | tags | recall@k | mrr | faithfulness | latency (ms) |");
  lines.push("| --- | --- | --- | --- | --- | --- |");
  for (const item of report.items) {
    lines.push(
      `| ${item.id} | ${item.tags.join(",")} | ${item.recallAtK.toFixed(2)} | ${item.mrr.toFixed(2)} | ${item.faithfulness.toFixed(2)} | ${item.latencyMs.toFixed(2)} |`,
    );
  }
  lines.push("");
  return lines.join("\n");
}

export function writeReport(report: EvalReport, outDir: string): void {
  fs.mkdirSync(outDir, { recursive: true });
  fs.writeFileSync(path.join(outDir, "results.json"), JSON.stringify(report, null, 2));
  fs.writeFileSync(path.join(outDir, "report.md"), reportMarkdown(report));
}
